From a9000b6724ae99a1b96baf6ab338cea9f0795629 Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Thu, 28 Jan 2021 12:04:32 +0800
Subject: [PATCH 29/75] v4l2videodec: Add hantro tiled output support

(cherry picked from commit e1899b66986b9c7bc2ede0e4a953bd37066900d8,
commit d3051d2ab731cea96b3060bbba6d0c89ec5becbe)

Upstream-Status: Pending

Signed-off-by: Hou Qi <qi.hou@nxp.com>

 Conflicts:
	sys/v4l2/gstv4l2bufferpool.c
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 sys/v4l2/ext/videodev2.h     |  2 +
 sys/v4l2/gstv4l2bufferpool.c | 10 ++++
 sys/v4l2/gstv4l2object.c     | 98 ++++++++++++++++++++++++++++++++++++
 sys/v4l2/gstv4l2object.h     |  2 +
 sys/v4l2/gstv4l2videodec.c   | 12 ++++-
 5 files changed, 122 insertions(+), 2 deletions(-)

diff --git a/sys/v4l2/ext/videodev2.h b/sys/v4l2/ext/videodev2.h
index 821c3b1..ba9bfe6 100644
--- a/sys/v4l2/ext/videodev2.h
+++ b/sys/v4l2/ext/videodev2.h
@@ -627,6 +627,8 @@ struct v4l2_pix_format {
 #define V4L2_PIX_FMT_P012M   v4l2_fourcc('P', 'M', '1', '2') /* 24  Y/CbCr 4:2:0 12-bit per component */
 #define V4L2_PIX_FMT_NV12X	 v4l2_fourcc('N', 'V', 'X', '2') /* Y/CbCr 4:2:0 for 10bit  */
 #define V4L2_PIX_FMT_NV12_10BIT   v4l2_fourcc('N', 'T', '1', '2') /* Y/CbCr 4:2:0 for 10bit  */
+#define V4L2_PIX_FMT_RFC	 v4l2_fourcc('R', 'F', 'C', '0') /* 8bit tile output, with rfc*/
+#define V4L2_PIX_FMT_RFCX	 v4l2_fourcc('R', 'F', 'C', 'X') /* 10 bit tile output, with rfc */
 
 /* three planes - Y Cb, Cr */
 #define V4L2_PIX_FMT_YUV410  v4l2_fourcc('Y', 'U', 'V', '9') /*  9  YUV 4:1:0     */
diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index 7bfb2b5..29f3d9f 100644
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -39,6 +39,8 @@
 #include "gst/video/gstvideometa.h"
 #include "gst/video/gstvideopool.h"
 #include "gst/allocators/gstdmabuf.h"
+#include "gst/allocators/gstphymemmeta.h"
+#include "gstimxcommon.h"
 
 #include <gstv4l2bufferpool.h>
 
@@ -1245,6 +1247,7 @@ gst_v4l2_buffer_pool_dqbuf (GstV4l2BufferPool * pool, GstBuffer ** buffer,
   GstClockTime timestamp;
   GstV4l2MemoryGroup *group;
   const GstVideoInfo *info = &obj->info;
+  GstPhyMemMeta *pmeta = { 0 };
   gint i;
   gint old_buffer_state;
 
@@ -1420,6 +1423,13 @@ gst_v4l2_buffer_pool_dqbuf (GstV4l2BufferPool * pool, GstBuffer ** buffer,
   GST_BUFFER_OFFSET (outbuf) = group->buffer.sequence;
   GST_BUFFER_OFFSET_END (outbuf) = group->buffer.sequence + 1;
 
+  if (IS_IMX8MQ () && obj->drm_modifier != 0
+      && !V4L2_TYPE_IS_OUTPUT (obj->type)) {
+    pmeta = GST_PHY_MEM_META_ADD (outbuf);
+    pmeta->rfc_luma_offset = group->buffer.reserved;
+    pmeta->rfc_chroma_offset = group->buffer.reserved2;
+  }
+
 done:
   *buffer = outbuf;
 
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index ef327ff..4347311 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -38,11 +38,14 @@
 #include "gstv4l2object.h"
 #include "gstv4l2tuner.h"
 #include "gstv4l2colorbalance.h"
+#include "gstimxcommon.h"
 
 #include <glib/gi18n-lib.h>
 
 #include <gst/video/video.h>
 #include <gst/allocators/gstdmabuf.h>
+#include <gst/allocators/gstdmabufmeta.h>
+#include <libdrm/drm_fourcc.h>
 
 GST_DEBUG_CATEGORY_EXTERN (v4l2_debug);
 #define GST_CAT_DEFAULT v4l2_debug
@@ -167,6 +170,8 @@ static const GstV4L2FormatDesc gst_v4l2_formats[] = {
   {MAP_FMT (NV12M_10BE_8L128, NV12_10BE_8L128), GST_V4L2_RAW},
   {MAP_FMT (NV12_10BIT, NV12_10LE40),           GST_V4L2_RAW},
   {MAP_FMT (NV12X, NV12_10LE40),                GST_V4L2_RAW},
+  {MAP_FMT (RFC, NV12),                         GST_V4L2_RAW},
+  {MAP_FMT (RFCX, NV12_10LE40),                 GST_V4L2_RAW},
   {MAP_FMT (NV21M, NV21),                       GST_V4L2_RAW},
   {MAP_FMT (NV21, NV21),                        GST_V4L2_RAW},
   {MAP_FMT (NV16M, NV16),                       GST_V4L2_RAW},
@@ -555,6 +560,9 @@ gst_v4l2_object_new (GstElement * element,
   v4l2object->channels = NULL;
   v4l2object->colors = NULL;
 
+  v4l2object->drm_modifier = 0;
+  v4l2object->is_g2 = FALSE;
+
   v4l2object->keep_aspect = TRUE;
 
   v4l2object->n_v4l2_planes = 0;
@@ -1127,6 +1135,8 @@ gst_v4l2_object_format_get_rank (const struct v4l2_fmtdesc *fmt)
       rank = GREY_BASE_RANK;
       break;
 
+    case V4L2_PIX_FMT_RFC:     /* 8-bit tile output  */
+    case V4L2_PIX_FMT_RFCX:    /* 10-bit tile output  */
     case V4L2_PIX_FMT_NV12MT:  /* NV12 64x32 tile   */
     case V4L2_PIX_FMT_NV21:    /* 12  Y/CrCb 4:2:0  */
     case V4L2_PIX_FMT_NV21M:   /* Same as NV21      */
@@ -3748,6 +3758,9 @@ gst_v4l2_object_set_format_full (GstV4l2Object * v4l2object, GstCaps * caps,
     goto invalid_caps;
 
   pixelformat = fmtdesc->pixelformat;
+  if (V4L2_TYPE_IS_OUTPUT (v4l2object->type)
+      && (pixelformat == V4L2_PIX_FMT_HEVC || pixelformat == V4L2_PIX_FMT_VP9))
+    v4l2object->is_g2 = TRUE;
   width = GST_VIDEO_INFO_WIDTH (&info);
   height = GST_VIDEO_INFO_FIELD_HEIGHT (&info);
   /* if caps has no width and height info, use default value */
@@ -5355,6 +5368,8 @@ gst_v4l2_object_decide_allocation (GstV4l2Object * obj, GstQuery * query)
   GstAllocator *allocator = NULL;
   GstAllocationParams params = { 0 };
   guint video_idx;
+  gboolean alloc_has_meta = FALSE;
+  guint alloc_index;
 
   GST_DEBUG_OBJECT (obj->dbg_obj, "decide allocation");
 
@@ -5388,6 +5403,89 @@ gst_v4l2_object_decide_allocation (GstV4l2Object * obj, GstQuery * query)
   GST_DEBUG_OBJECT (obj->dbg_obj, "allocation: size:%u min:%u max:%u pool:%"
       GST_PTR_FORMAT, size, min, max, pool);
 
+  alloc_has_meta =
+      gst_query_find_allocation_meta (query, GST_DMABUF_META_API_TYPE,
+      &alloc_index);
+
+  if (IS_IMX8MQ () || IS_AMPHION ()) {
+    if (alloc_has_meta) {
+      const GstStructure *params;
+      gint j, len;
+
+      gst_query_parse_nth_allocation_meta (query, alloc_index, &params);
+      GST_DEBUG_OBJECT (obj->dbg_obj,
+          "Expected field 'GstDmabufMeta' in structure: %" GST_PTR_FORMAT,
+          params);
+      if (params) {
+        const GValue *vdrm_modifier =
+            gst_structure_get_value (params, "dmabuf.drm_modifier");
+        gchar *meta = gst_structure_to_string (params);
+        if (GST_VALUE_HOLDS_LIST (vdrm_modifier)) {
+          len = gst_value_list_get_size (vdrm_modifier);
+          for (j = 0; j < len; j++) {
+            const GValue *val;
+            val = gst_value_list_get_value (vdrm_modifier, j);
+            guint64 drm_modifier = g_value_get_uint64 (val);
+            GST_DEBUG_OBJECT (obj->dbg_obj,
+                "dmabuf meta has modifier: %" G_GUINT64_FORMAT, drm_modifier);
+            if (IS_AMPHION () && drm_modifier == DRM_FORMAT_MOD_AMPHION_TILED) {
+              obj->drm_modifier = drm_modifier;
+            } else if (IS_IMX8MQ ()
+                && drm_modifier == DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED
+                && obj->is_g2 == TRUE) {
+              obj->drm_modifier = drm_modifier;
+              if (obj->format.fmt.pix.pixelformat == V4L2_PIX_FMT_NV12X)
+                obj->format.fmt.pix.pixelformat = V4L2_PIX_FMT_RFCX;
+              /* Hdr10 video does renegotiation to transfer hdr10 metadata. Capture pixelformat has been
+               * changed from NV12X to RFCX at the first time, so just keep the RFCX format here. */
+              else if (obj->format.fmt.pix.pixelformat != V4L2_PIX_FMT_RFCX)
+                obj->format.fmt.pix.pixelformat = V4L2_PIX_FMT_RFC;
+            } else {
+              GST_WARNING_OBJECT (obj->dbg_obj,
+                  "video sink can't support modifier: %lld",
+                  DRM_FORMAT_MOD_AMPHION_TILED);
+            }
+          }
+        } else if (meta) {
+          guint64 drm_modifier;
+          GST_DEBUG_OBJECT (obj->dbg_obj, "dmabuf meta has modifier: %s", meta);
+          sscanf (meta,
+              "GstDmabufMeta, dmabuf.drm_modifier=(guint64){ %" G_GUINT64_FORMAT
+              " };", &drm_modifier);
+          GST_DEBUG_OBJECT (obj->dbg_obj,
+              "dmabuf meta has modifier: %" G_GUINT64_FORMAT, drm_modifier);
+          if (IS_AMPHION () && drm_modifier == DRM_FORMAT_MOD_AMPHION_TILED) {
+            GST_DEBUG_OBJECT (obj->dbg_obj,
+                "video sink support modifier: %" G_GUINT64_FORMAT,
+                drm_modifier);
+            obj->drm_modifier = drm_modifier;
+          } else if (IS_IMX8MQ ()
+              && drm_modifier == DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED
+              && obj->is_g2 == TRUE) {
+            GST_DEBUG_OBJECT (obj->dbg_obj,
+                "video sink support modifier: %" G_GUINT64_FORMAT,
+                drm_modifier);
+            obj->drm_modifier = drm_modifier;
+            if (obj->format.fmt.pix.pixelformat == V4L2_PIX_FMT_NV12X)
+              obj->format.fmt.pix.pixelformat = V4L2_PIX_FMT_RFCX;
+            else if (obj->format.fmt.pix.pixelformat != V4L2_PIX_FMT_RFCX)
+              obj->format.fmt.pix.pixelformat = V4L2_PIX_FMT_RFC;
+          } else {
+            GST_WARNING_OBJECT (obj->dbg_obj,
+                "video sink can't support modifier: %lld",
+                DRM_FORMAT_MOD_AMPHION_TILED);
+          }
+        }
+        if (obj->ioctl (obj->video_fd, VIDIOC_S_FMT, &obj->format) < 0)
+          GST_WARNING_OBJECT (obj->dbg_obj, "VIDIOC_S_FMT failed");
+        else
+          GST_DEBUG_OBJECT (obj->dbg_obj,
+              "Set tiled output format %" GST_FOURCC_FORMAT,
+              GST_FOURCC_ARGS (obj->format.fmt.pix.pixelformat));
+      }
+    }
+  }
+
   has_video_meta =
       gst_query_find_allocation_meta (query, GST_VIDEO_META_API_TYPE,
       &video_idx);
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index a3173ba..85e591a 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -154,6 +154,8 @@ struct _GstV4l2Object {
   /* Features */
   gboolean need_video_meta;
   gboolean has_alpha_component;
+  guint64 drm_modifier;
+  gboolean is_g2;
 
   /* only used if the device supports MPLANE
    * nb planes is meaning of v4l2 planes
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 7b60989..1fb7316 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -429,6 +429,8 @@ gst_v4l2_video_dec_negotiate (GstVideoDecoder * decoder)
   self->v4l2capture->info.fps_d = self->v4l2output->info.fps_d;
   self->v4l2capture->info.fps_n = self->v4l2output->info.fps_n;
 
+  self->v4l2capture->is_g2 = self->v4l2output->is_g2;
+
   /* For decoders G_FMT returns coded size, G_SELECTION returns visible size
    * in the compose rectangle. gst_v4l2_object_acquire_format() checks both
    * and returns the visible size as with/height and the coded size as
@@ -886,10 +888,16 @@ gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
 
     frame->duration = self->v4l2capture->duration;
     frame->output_buffer = buffer;
-    if (IS_AMPHION () && self->v4l2capture->is_amphion) {
+    if (IS_IMX8MQ () && self->v4l2capture->is_hantro) {
+      guint64 drm_modifier = self->v4l2capture->drm_modifier;
+      gst_buffer_add_dmabuf_meta (frame->output_buffer, drm_modifier);
+      GST_DEBUG_OBJECT (decoder, "Add drm modifier: %" G_GUINT64_FORMAT,
+          drm_modifier);
+    } else if (IS_AMPHION () && self->v4l2capture->is_amphion) {
       guint64 drm_modifier = DRM_FORMAT_MOD_AMPHION_TILED;
       gst_buffer_add_dmabuf_meta (frame->output_buffer, drm_modifier);
-      GST_DEBUG_OBJECT (decoder, "Add drm modifier: %lld\n", drm_modifier);
+      GST_DEBUG_OBJECT (decoder, "Add drm modifier: %" G_GUINT64_FORMAT,
+          drm_modifier);
     }
     buffer = NULL;
     ret = gst_video_decoder_finish_frame (decoder, frame);
-- 
2.43.0

